orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/util/PackageInfo.groovy

```diff
@@ -215,4 +228,20 @@ class PackageInfo {
     }
   }
 
+  @CompileDynamic
+  private Map findBuildInfoInUpstreamStage() {
+
+    stage?.ancestors()?.findAll {
+      it?.stage?.context?.buildInfo?.artifacts
+    }?.find {
+      artifactMatch(it.stage.context?.buildInfo?.artifacts as List)
+    }?.stage?.context?.buildInfo as Map
+  }
+
+  @CompileDynamic
+  private boolean artifactMatch(final List artifacts) {
```

Comment:
We don't generally apply `final` to parameters unless necessary ... although I'd probably suggest a stronger type `List<Map<String, String>>`.

That might actually let you get rid of the `@CompileDynamic`.